prototypes/process-explorer/dotnet/src/MorganStanley.ComposeUI.ProcessExplorer.Abstractions/Entities/ConnectionInfo.cs

```diff
@@ -0,0 +1,98 @@
+
+public sealed class ConnectionInfo : IConnectionInfo
+{
+    public ConnectionInfo(
+        Guid id,
+        string name,
+        ConnectionStatus status,
+        string? localEndpoint = null,
+        string? remoteEndpoint = null,
+        string? remoteApplication = null,
+        string? remoteHostname = null,
+        IEnumerable<KeyValuePair<string, string>>? connectionInformation = null)
+    {
+        Id = id;
+        Name = name;
+        LocalEndpoint = localEndpoint;
+        RemoteEndpoint = remoteEndpoint;
+        RemoteApplication = remoteApplication;
+        RemoteHostname = remoteHostname;
+
+        if (connectionInformation != null)
+            ConnectionInformation = new(connectionInformation);
+
+        Status = status.ToStringCached();
+    }
+
+    public Guid Id { get; init; }
+    public string Name { get; init; }
+    public string? LocalEndpoint { get; private set; }
+    public string? RemoteEndpoint { get; private set; }
+    public string? RemoteApplication { get; private set; }
+    public string? RemoteHostname { get; private set; }
+    public ConcurrentDictionary<string, string>? ConnectionInformation { get; init; }
+    public string Status { get; private set; }
+    protected readonly Subject<KeyValuePair<string, ConnectionStatus>> ConnectionStatusSubject = new();
+    public IObservable<KeyValuePair<string, ConnectionStatus>> ConnectionStatusEvents => ConnectionStatusSubject;
+
+    private readonly object _connectionInformationLock = new();
+
+    public void UpdateConnection(
+        string? localEndpoint = null,
+        string? remoteEndpoint = null,
+        string? remoteApplication = null,
+        string? remoteHostname = null,
+        IEnumerable<KeyValuePair<string, string>>? connectionInformation = null,
+        ConnectionStatus status = ConnectionStatus.Unknown)
+    {
+        if (localEndpoint != null) LocalEndpoint = localEndpoint;
+        if (remoteEndpoint != null) RemoteEndpoint = remoteEndpoint;
+        if (remoteApplication != null) RemoteApplication = remoteApplication;
+        if (remoteHostname != null) RemoteHostname = remoteHostname;
+        if (connectionInformation != null && connectionInformation.Any()) AddOrUpdateConnectionInformation(connectionInformation);
+
+        if (status != ConnectionStatus.Unknown)
+        {
+            Status = status.ToStringCached();
+            ConnectionStatusSubject.OnNext(new(Id.ToString(), status));
+        }
+    }
+
+    private void AddOrUpdateConnectionInformation(IEnumerable<KeyValuePair<string, string>> connectionInformation)
+    {
+        lock (_connectionInformationLock)
+        {
+            foreach (var connectionInfo in connectionInformation)
+            {
+                ConnectionInformation.AddOrUpdate(connectionInfo.Key, connectionInfo.Value,
+                    (_, _) => connectionInfo.Value);
+            }
+        }
+    }
+
+    public override bool Equals(object? obj)
```

Comment:
"Types that override [Equals(Object)](https://learn.microsoft.com/en-us/dotnet/api/system.object.equals?view=net-7.0#system-object-equals(system-object)) must also override [GetHashCode](https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-7.0); otherwise, hash tables might not work correctly."

Reply:
You are right, I totally missed that :)